### DISCUSSION TITLE
Questions on Applications Software Architecture

### DISCUSSION THREAD

**Discussion by Abishtu** [organicmaps/organicmaps#5065] [Q&A] [answered]
Hello, I've been spending a couple days looking through the code trying to figure out how the application works and what it's architecture would look like, right now I'm guessing it's something like a multi-platform front-end that interfaces with a common C++ core for the back-end, with some platforms like android using the NDK as a wrapper for the C++ code?

I now have a couple question on how data is communicated around. From the digging I've done thus far, I'm guessing that you guys are using REST API requests to communicate b/w the interface and the "back-end" for things like searching for locations, bookmarks, finding routes and other map data including rendering?

With regards to how the routing works, I can see that it's a mix of OSRM and a in-house routing systems, do you have an OSRM http server running locally per device that you communicate with using the [OSRM api](http://project-osrm.org/docs/v5.5.1/api/#general-options) and then translate back to whatever would work for the mwm format being used?

Finally how is the mwm data being read? I am aware the `mwm` files are compressed version of OSM map data used, and there are translation files b/w `mwm` and OSM tags, so do you decompress the data when it's being used and then re-compress it back to save space whilst idle?

**Comment by biodranik** [comment, marked as answer]
1. We don't use REST API. C++ is our API.
2. We don't use OSRM anymore for many years.
3. mwm files are created by a generator and only read by a client. You can dig deeper to read the code.

What is your goal? Why are you studying the code?

**Reply by Abishtu** [reply]
Its part of an assignment for university software architecture course, CSSE6400 ([https://github.com/CSSE6400](https://github.com/CSSE6400)), its an open source course so the course website and content are open for anyone to access. We're tasked with finding a complex open source project and finding out how it works, documenting things like its Architecturally Significant Requirements (ASRs), and diagrams of what its Software Architecture looks like.

(Also cause I wish to contribute to this project in the future, so studying the code as a uni assignment acts as a strong motivator.)

**Reply by barisunsalhn** [reply]
About 2. Are you using bidirectional a* (as you mentioned in another discussion )?

**Reply by biodranik** [reply]
Yes.
